nncf/common/tensor.py

```diff
@@ -8,40 +8,358 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+import abc
 from abc import abstractmethod
-from typing import List, Optional, TypeVar
+from enum import IntEnum
+from typing import Any, Callable, Generic, List, Optional, Tuple, Type, TypeVar, Union
+from typing import Iterator
+
+import numpy as np
 
 TensorType = TypeVar("TensorType")
 DeviceType = TypeVar("DeviceType")
-TensorElementsType = TypeVar("TensorElementsType")
 
 
-class NNCFTensor:
+class TensorDtype(IntEnum):
+    FLOAT32 = 0
+    INT64 = 1
+
+
+class NNCFTensor(Generic[TensorType], abc.ABC):
```

Comment:
Backend implementations only need to implement this class and the `NNCFTensorBackend`. The methods in both classes are supposed to mirror the `numpy` syntax in terms of meaning, method names, parameter names, default values, so that the developers could intuitively use numpy-like reasoning and get consistent results.